### ISSUE TITLE
ngcc: is unable to process some minified library formats

### ISSUE THREAD

**Issue by ARTWINPRO** [angular/angular#35255]
**Error with old package.**

Fix please.

![image](https://user-images.githubusercontent.com/16427316/74086971-df47eb80-4a90-11ea-9be8-59f851ec40d6.png)

**Comment by JoostK** [comment]
The `daypilot-pro-angular` library is published in a minified format, which cannot be compiled for Ivy consumption by ngcc (the compatibility compiler) as its contents are not fully statically analyzable.

I'm afraid we can't really make this work, without the publisher redistributing the packages in a non-minified format. Please refer to the thread in #32442 for more information on the various minifier options that ngcc is/is not able to work with.

**Comment by ARTWINPRO** [comment]
@JoostK maybe i can include min,js file in angular-cli.json ?

But if yes, how will be with min.d.ts file ?

**Comment by JoostK** [comment]
I'm sorry, I don't understand what you're asking about.

**Comment by ARTWINPRO** [comment]
Srry but maybe you have some resolve for minify packages ?

**Comment by mevlts** [comment]
I also have the same issue for package "ng2-currency mask" and couldn't find any suggestion.

**Comment by ARTWINPRO** [comment]
Guys, i have some solution but im not sure how currect it is.

```
1. "buildOptimizer": false,
2. "enableIvy": true
3. "aot": false,

```

Before do this refactoring, application size is: 18 440KB;
After: 17 887KB

**Comment by roopeshreddy** [comment]
I've created a library repo here - [Angular Library project](https://github.com/roopeshreddy/angular-lib-demo)

OR

Install this - **ngx-lib-ivy-bug** and do **ng serve** in any blank Angular v9.x project with "aot": true.

Thanks

**Comment by air-Design** [comment]
I am facing similar issues after updating again from v8 to v9 need help please
![ngcc error](https://user-images.githubusercontent.com/28719431/74272073-e9265480-4d05-11ea-8ccd-c075c3718e5a.JPG)

**Comment by VedantKarpe** [comment]
Hi Guys,

I tried to install Angular Material 9 but its also failing with the same error.

![image](https://user-images.githubusercontent.com/29578397/74330886-86da5c00-4db8-11ea-82d2-9034d485bf1c.png)

**Comment by brianpilati** [comment]
I am seeing a similar problem on an imported library:

Failed to compile entry-point x-x-x (es2015 as esm2015) due to compilation errors:
node_modules/ngx-socket-io/src/socket-io.module.d.ts:7:22 - error NG6002: Appears in the NgModule.imports of x-x-x, but could not be resolved to an NgModule class

I'm going to have to either disable Ivy or role back my installation.

**Comment by SoterRamirez** [comment]
My temporary solution was to add:
  **"enableIvy": false,** 

`"angularCompilerOptions": {`
   ` "fullTemplateTypeCheck": true,`
    ` "enableIvy": false, <---`
  `  "strictInjectionParameters": true`
  ` }`

from my file  _tsconfig.json_

**Comment by brianpilati** [comment]
Once I displayed Ivy, same as @SoterRamirez, my build worked again. Also, when building a library for an NPM publish, there was a warning to "NOT use ivy when publishing NPMs". So at this point, unless someone shares a new thread -- IVY IS NOT available for building ANY library packaged with anything less than 9.

**Comment by gkalpak** [comment]
Published libraries should not be published built with Ivy, because (a) that would prevent non-Ivy apps to consume them and (b) the Ivy instructions are still not considered public API and may break in the future.

That said, you should be able to consume any libraries in your Ivy app. To support that, there is a compatibility compiler (ngcc) that will process all dependencies of your app and convert them to Ivy. (That's where the `Compiling entry-point XYZ (<format-property> as <format>)` are coming from.)

The error you are seeing mean that ngcc was not able to process a particular library. Please, share a minimal reproduction of the problem (not just the error message), so we can investigate (and hopefully fix) the problem 🙏

**Comment by PradipShrestha** [comment]
Getting it for HttpClientModule. Only thing, I did was add that module in app.module.
```` bash
ERROR in ../../node_modules/@angular/common/http/http.d.ts:2801:22 - error NG6002: Appears in the NgModule.imports of AppModule, but could not be resolved to an NgModule class
    
    2801 export declare class HttpClientModule {
                              ~~~~~~~~~~~~~~~~

````

**Comment by ssuperczynski** [comment]
@gkalpak I am able to reproduce this error:

1. There is a demo project https://github.com/ssuperczynski/ngx-easy-table with library `ngx-easy-table` in the projects dir.
2. I updated angular to 9 by `ng update @angular/cli` `ng update angular/core`
3. Library is compiled by ng-packagr using command `ng build ngx-easy-table`
4. IVY is enabled in the demo (tsconfig.json), but disabled in the library (projects/ngx-easy-table/tsconfig.lib.json).
5. Once step 2 is finished I build demo app `ng build demo --prod --configuration=production`

Error appears at the end of the 5th step:

```

ERROR in dist/ngx-easy-table/lib/table.module.d.ts:1:22 - error NG6002: Appears in the NgModule.imports of AppModule, but could not be resolved to an NgModule class

1 export declare class TableModule {
                       ~~~~~~~~~~~

npm ERR! code ELIFECYCLE
```

Below link to PR and travis where this error appears:
PR: https://github.com/ssuperczynski/ngx-easy-table/pull/285
Travis: https://travis-ci.org/ssuperczynski/ngx-easy-table/jobs/650322357?utm_medium=notification&utm_source=github_status

**Comment by houdass** [comment]
Hi guys, I figured out what going wrong, I had a Angular 8 project (containing a library project) that I have upgraded to Angular 9 (with a lib also: ng generate library core, I created at first a new angular 9 project to compare the differences with my project made with ng 8, i figure out that there is some things missing, here is the steps I followed:
- Delete node_modules
- Delete dist folder
- In the Angular 9 project lib there is a new file called: tsconfig.lib.prod.json 
```
{
  "extends": "./tsconfig.lib.json",
  "angularCompilerOptions": {
    "enableIvy": false
  }
}
```
- In your angular.json add the following:
```
{
  ...
  "projects": {
     ...
     "configurations": {
            "production": {
              "tsConfig": "projects/core/tsconfig.lib.prod.json"
             }
      }
  ...
  }
}
```
- run ng build --prod, and everything will work :D

PS: in this case the library is called "core"

**Comment by rbelmega** [comment]
@houdass you are just setting 
`"enableIvy": false`
for the main project, which has been mentioned before by @SoterRamirez
for a temporary workaround this should be fine, but for **temporary**

**Comment by PritiMaurya** [comment]
I also update angular 8 to 9 I am getting this error at the time of build

![Screenshot 2020-02-14 14 50 00](https://user-images.githubusercontent.com/35628359/74518334-a51b9580-4f39-11ea-9697-260561dd1835.png)

**Comment by gkalpak** [comment]
@ssuperczynski, can you please open a new issue (with the info in https://github.com/angular/angular/issues/35255#issuecomment-586138300), as it seems to be a different problem?

**Comment by danrevah** [comment]
Well, I did try to upgrade `ngx-pipes` to support Angular 9, but was encountering the same error while generating a fresh new Angular project (ng new ...).

Also, while trying to disable Ivy, it didn't work either.
with a new error:
```
ERROR in Unexpected value 'NgPipesModule in /Users/xyz/tmp2/angular-test/node_modules/ngx-pipes/src/ng-pipes/ng-pipes.module.d.ts' imported by the module 'AppModule in /Users/xyz/tmp2/angular-test/src/app/app.module.ts'. Please add a @NgModule annotation.
```

**Comment by arthurflor23** [comment]
Same issue with completer and smart-table packages

**Comment by CharlyRipp** [comment]
Hi, figured a simplest recreation would help. Apologies in advance if it isn't :)

Here's how to replicate via script.

```
tmpDir=$(pwd)
ng new lib-issue --create-application=false --minimal
cd lib-issue
ng g lib my-lib
ng build --project=my-lib --prod
cd dist/my-lib
npm link

cd $tmpDir
ng new lib-issue-app --minimal --routing --interactive=false
cd lib-issue-app
npm link my-lib

## Import the library module
cat << "EOF" > src/app/app.module.ts
import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';

import { AppRoutingModule } from './app-routing.module';
import { AppComponent } from './app.component';
import { MyLibModule } from 'my-lib';

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    AppRoutingModule,
    MyLibModule,
  ],
  providers: [],
  bootstrap: [AppComponent]
})
export class AppModule { }
EOF

## Use the library component
cat << "EOF" > src/app/app.component.ts
import { Component } from '@angular/core';

@Component({
  selector: 'app-root',
  template: `<lib-my-lib></lib-my-lib>`,
  styles: []
})
export class AppComponent {}
EOF

## !! Manual !!
## Update angular.json to preserve symlinks via npm link
## "projects": {
##   "lib-issue-app": {
##     "architect": {
##       "build": {
##         "options": {
##           "preserveSymlinks": true

npm run start
```
The above will produce a workspace w/ library, compile it via `--prod` as to produce it without Ivy (as per warned), then produce a workspace with an application and `npm link` the library in.

Results in:
```
ERROR Error: Uncaught (in promise): TypeError: Cannot read property 'bindingStartIndex' of null
TypeError: Cannot read property 'bindingStartIndex' of null
    at Module.ɵɵelementStart (core.js:20925)
    at MyLibComponent_Template (my-lib.js:19)
    at executeTemplate (core.js:11858)
```

When turning off Ivy ("enableIvy": false --> tsconfig.json) in the application, it runs fine.

As for my personal use-case, I'm wanting to do the 'proxy' method to lazy load components from the library, which results in an identical trace. Also works with Ivy disabled.

My personal workaround is going to be to disable ivy on my applications for now until a solution is reached.

**Comment by Rudedog9d** [comment]
Likely unrelated to _most_ people's issues listed here, but I had the same error displayed while working under a symlink on Windows. Moving to a "real" path solved the issue. (Related issue https://github.com/angular/angular-cli/issues/9909)

I got here after by googling the same error as OP (but with different package), so atleast somewhat relevant 🤷‍♂️

**Comment by CharlyRipp** [comment]
Under symlink on osx, found that I had to rebuild the application twice. Produces a `__ivy_ngcc__` on the first build, which does seem to be in time? Then restarting the dev server seems to pick it up fine.
Haven't experimented with builds or non-symlinks yet.

**Comment by pklejnowski** [comment]
Is there any way to make it work with `aot=true`
I have the same problem in my Hybrid application
![image](https://user-images.githubusercontent.com/28590512/74936947-654a2780-53eb-11ea-9a69-f25b69483b59.png)

**Comment by DarkLite1** [comment]
I',m a total newbie at this and was following [this tutorial](https://youtu.be/1tRLveSyNz8?t=4930) when I got stuck. Mat said to add this line to `app.module.ts`:

`import { MatInputModule } from "@angular/material";`

This failed with the error:

> ERROR in src/app/app.module.ts:5:32 - error TS2306: File 'T:/acadamind/mean-course/node_modules/@angular/material/index.d.ts' is not a module.
> 
> 5 import { MatInputModule } from "@angular/material";

Trying to solve this I checked the [official documentation](https://material.angular.io/components/input/api) which states to use:

`import {MatInputModule} from '@angular/material/input'`

This seems to work fine. But when I try the following:
`import { MatInputModule, MatCardModule } from "@angular/material";`

it fails so I have to stick to using:
```javascript
import { MatInputModule } from "@angular/material/input";
import { MatCardModule } from "@angular/material/card";
```

I don't know if this is an Angular specific issue or that is iabout node.js importing modules. Using node.js `13.9.0`.

**Comment by gkalpak** [comment]
Importing from `@angular/material` is no longer supported. This is a breaking change in 9.0.0-next.0 (see [changelog][1]):

> Components can no longer be imported through "@angular/material". Use the individual secondary entry-points, such as `@angular/material/button`.

[1]: https://github.com/angular/components/blob/a1bfc4e2a/CHANGELOG.md#900-next0-cardboard-cpu-2019-10-03

**Comment by erickjhorman** [comment]
> 
> 
> I am facing similar issues after updating again from v8 to v9 need help please
> ![ngcc error](https://user-images.githubusercontent.com/28719431/74272073-e9265480-4d05-11ea-8ccd-c075c3718e5a.JPG)

You only must add the right API reference of the component you wanna use for example 
To add a toolbar you need import MatToolbarModule in this case

**Comment by QPeiran** [comment]
Every time I import a new angular material module it shows. However, it only appear once. If I restart server then this warning/error message gone. Just feel a bit disturbing..

**Comment by kumaresan-subramani** [comment]
> Hi guys, I figured out what going wrong, I had a Angular 8 project (containing a library project) that I have upgraded to Angular 9 (with a lib also: ng generate library core, I created at first a new angular 9 project to compare the differences with my project made with ng 8, i figure out that there is some things missing, here is the steps I followed:
> 
> * Delete node_modules
> * Delete dist folder
> * In the Angular 9 project lib there is a new file called: tsconfig.lib.prod.json
> 
> ```
> {
>   "extends": "./tsconfig.lib.json",
>   "angularCompilerOptions": {
>     "enableIvy": false
>   }
> }
> ```
> 
> * In your angular.json add the following:
> 
> ```
> {
>   ...
>   "projects": {
>      ...
>      "configurations": {
>             "production": {
>               "tsConfig": "projects/core/tsconfig.lib.prod.json"
>              }
>       }
>   ...
>   }
> }
> ```
> 
> * run ng build --prod, and everything will work :D
> 
> PS: in this case the library is called "core"

@houdass , **in here you are disabled IVY so that it will work**

**Comment by brianpilati** [comment]
Any updates on this issue. It's still an issue.

**Comment by kumaresan-subramani** [comment]
For me, while removing cache it works

**Comment by pklejnowski** [comment]
Adding `"postinstall": "ngcc"` to "scripts: {...}" in package.json solved problem.

**Comment by rsawik** [comment]
"enableIvy": false

 "dependencies": {
    "@angular/animations": "~9.0.4",
    "@angular/common": "~9.0.4",
    "@angular/compiler": "~9.0.4",
    "@angular/core": "~9.0.4",
    "@angular/forms": "~9.0.4",
    "@angular/localize": "^9.0.4",
    "@angular/platform-browser": "~9.0.4",
    "@angular/platform-browser-dynamic": "~9.0.4",
    "@angular/router": "~9.0.4",
    "@coreui/angular": "^2.9.0",
    "@coreui/coreui": "^2.1.16",
    "@ng-bootstrap/ng-bootstrap": "^5.3.0",
    "@ngx-loading-bar/core": "^4.2.0",
    "@ngx-loading-bar/http-client": "^4.2.0",
    "class-transformer": "^0.2.3"
  },
 "importHelpers": true, in tsconfig.json 

disable dotnet watch run 
ng build ng serve 

vuala

**Comment by pitrackster** [comment]
Hi,
Just faced the issue.
For me the solution found [here](https://github.com/angular/angular/issues/34983#issuecomment-586554284) worked perfectly

**Comment by shailugit** [comment]
Issue fixed after Adding "postinstall": "ngcc" to "scripts: {...}" in package.json

**Comment by Prowillz** [comment]
Simply remove whatever you're importing from "imports" and put it in "declarations" and then "exports"

**Comment by AkashGupta13** [comment]
**Deleting node modules** folder and doing **npm install** worked for me!

**Comment by lak-coder** [comment]
I upgraded to Angular 9, with all the dependencies also upgraded, I am getting a lot of errors and my app in not working anymore.

:: error NG6002: Appears in the NgModule.imports of OutModule, but could not beresolved to an NgModule class

For Temporary solution disabled Ivy in tsconfig.json:
-------------------------------------------------
"angularCompilerOptions": {
    "fullTemplateTypeCheck": true,
    "enableIvy": false,
    "strictInjectionParameters": true
  } 

It is working fine but when I would can use Ivy.

Is any solution for Angular9 Ivy?

**Comment by bhuber2010** [comment]
- Works fine in 9.0.7
- Do `ng update @angular/cli @angular/core` (to 9.1.0)
- Get the NG6002 error from postinstall on some libraries in my Nx monorepo

So seems there is a breaking change from 9.07 >> 9.1.0

**Comment by JoostK** [comment]
@bhuber2010 Could you please try updating to 9.1.1, there were a lot of ngcc fixes in that release.

**Comment by bhuber2010** [comment]
@JoostK Nope, same thing with 9.1.1

**Comment by JoostK** [comment]
@bhuber2010 okay, that is unexpected. Can you open a new issue with repro details as it's likely a different root cause than reported here.

Or maybe what you could try first is wiping out all of `node_modules` and installing again from scratch, we have seen cases where this helped.

**Comment by bhuber2010** [comment]
I fixed this by removing my local monorepo libs that I also have installed from our private npm (as a sanity test). Not totally sure why it worked fine like that from ~8 up to 9.0.7 but breaks on 9.1.0+. Maybe if I deployed new versions of the libs, it would work, not really sure though.

**Comment by RUPALAVECINO** [comment]
Lo estas importando mal

**Comment by flash-me** [comment]
Hi @cory-colt,
are you Importing the Datepicker module in your app? 

cheers
flash :zap:

**Comment by UmeshGEM** [comment]
ERROR in node_modules/ngx-qrcode2/index.d.ts:3:22 - error NG6002: Appears in the NgModule.imports of AppModule, but could not be resolved to an NgModule class.


I am getting this is error

**Comment by flash-me** [comment]
> 5\. ng build demo --prod --configuration=production

Followed the steps and successfully built it. Am I missing sth?
![image](https://user-images.githubusercontent.com/5240450/80027244-bfe32b00-84e3-11ea-8d40-968f8c6720a4.png)
 
cheers
flash :zap:

**Comment by flash-me** [comment]
> I've created a library repo here - [Angular Library project](https://github.com/roopeshreddy/angular-lib-demo)
> 
> OR
> 
> Install this - **ngx-lib-ivy-bug** and do **ng serve** in any blank Angular v9.x project with "aot": true.
> 
> Thanks

@roopeshreddy 

Did it. Works. Am I missing sth?
![image](https://user-images.githubusercontent.com/5240450/80028510-84496080-84e5-11ea-9201-28c0ecf09111.png)

cheers
flash :zap:

**Comment by cory-colt** [comment]
> Hi @cory-colt,
> are you Importing the Datepicker module in your app?
> 
> cheers
> flash ⚡

@flash-me yes, I am importing it. It seems to only be the Datepicker that I'm having this issue with at this point, which is a bit odd to me.

**Comment by flash-me** [comment]
To everyone else: I you want to get help by the community or even the angular team itself, it is not enough that you are encountered an issue. Some details to reproduce or even better a git repo would be very helpful.

https://github.com/angular/angular/blob/master/CONTRIBUTING.md#submit-issue

cheers
flash :zap:

**Comment by flash-me** [comment]
> yes, I am importing it. It seems to only be the Datepicker that I'm having this issue with at this point, which is a bit odd to me.

@cory-colt 

Cannot reproduce. Steps I did
1. Generate demo application with cli: `ng new --commit false --minimal -s -t -g --S --style scss foo`
2. Added material: `npm i @angular/{cdk,material}`
3. Added  `BrowserAnimationsModule, MatNativeDateModule, MatDatepickerModule, MatInputModule`, to the `app.module.ts`
4. Copied the [first sample](https://material.angular.io/components/datepicker/examples) into the template of the `app.component.ts`
5. Works

May you provide a git repo?

cheers
flash :zap:

**Comment by jppgit** [comment]
> 
> 
> Hi,
> Just faced the issue.
> For me the solution found [here](https://github.com/angular/angular/issues/34983#issuecomment-586554284) worked perfectly

Mis perros, i add just this i dont know why but if you use ng add material or ng add firebase this is generating that issue for me,  just use normal npm install, i deleted my node_modules, do npm installl and all works like a charm again over angular 9 project, seeee seeee : )

**Comment by mrblack360** [comment]
Just restart the `npm` in order to reload the new changes you have made in the `app.module.ts`.

**Comment by lethuydung0109** [comment]
> Just restart the `npm` in order to reload the new changes you have made in the `app.module.ts`.

It works like a charm ! 

'npm restart' does the trick !

**Comment by petebacondarwin** [comment]
I have changed the title of this issue since the error message shown is too vague and captures a number of different possible issues, which is why this issue has become a bit of a catch-all for lots of messages. If you issue is not related to minified library code then please open a new issue.

**Comment by Nandakyawwin** [comment]
Guy!! If you are linux user,you need "root" access. At first , type => sudo su and type your password.
And then try again => ng serve .

**Comment by flash-me** [comment]
> Guy!! If you are linux user,you need "root" access. At first , type => sudo su and type your password.
> And then try again => ng serve .

If you need root access to compile your angular app you probably have a completely wrong nodejs/npm configuration & installed your node_modules etc. in restricted directories.

For anyone else: **NO. JUST NO. SERIOUSLY. Avoid using sudo for your daily workflow as much as possible...because security.**

cheer
flash ⚡

**Comment by admosity** [comment]
Hopefully this will be of help to someone to resolve this issue. We have an internal ui package built with ng-packagr@9.1.5. It is a fairly large library and upon hearing the benefits of compilation times we hopped on typescript@3.9.x. It seems that the output of the build is vastly different between the two versions. 

The diffs are from a non working build to a working build:

This is seen throughout most fields:

![image](https://user-images.githubusercontent.com/4655972/83445067-51ff1d00-a41a-11ea-8275-9b135326fef9.png)

The module declaration code is vastly different:

![image](https://user-images.githubusercontent.com/4655972/83445556-30eafc00-a41b-11ea-9ce3-328898aadb8a.png)

Consumers of the library had what most others seen here:
![image](https://user-images.githubusercontent.com/4655972/83445843-bcfd2380-a41b-11ea-950a-26be19336e57.png)

**Comment by petebacondarwin** [comment]
@admosity - thanks for posting this. Yes TS 3.9 brought in a number of painful breaking changes. The version of Angular that supports TS 3.9 will be 10.0.0 - so do not expect Angular 9.x to work with TS 3.9.

**Comment by mittulmandhan** [comment]
replace "target": "es5" with "target": "es2015" at src/e2e/tsconfig.json

**Comment by petebacondarwin** [comment]
There is a limited amount that ngcc can do to decipher packages that are heavily minified/obfuscated. There is actually very little benefit to most (or maybe even all) libraries by trying to obfuscate the distributed code.
We will attempt to fix issues with such libraries on a case by case basis if it can be done in a reasonable manner. But there is little point in keeping this general issue open.

**Comment by atwright147** [comment]
Given all of the above, how would I stop `ng-packagr` from minifying my own Angular lib?
